mat64/io.go

```diff
@@ -5,14 +5,171 @@
+
+	"github.com/gonum/matrix"
 )
 
-var (
-	littleEndian  = binary.LittleEndian
-	bigEndian     = binary.BigEndian
-	defaultEndian = littleEndian
+// maxLen is the biggest slice/array len one can create on a 32/64b platform.
+const maxLen = int64(int(^uint(0) >> 1))
 
-	sizeInt64   = binary.Size(int64(0))
-	sizeFloat64 = binary.Size(float64(0))
+var (
+	sizeInt64     = binary.Size(int64(0))
+	sizeFloat64   = binary.Size(float64(0))
+	errDataTooBig = errors.New("mat64: resulting data slice too big")
 )
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Dense is little-endian encoded as follows:
+//   0 -  7  number of rows    (int64)
+//   8 - 15  number of columns (int64)
+//  16 - ..  matrix data elements (float64)
+//           [0,0] [0,1] ... [0,ncols-1]
+//           [1,0] [1,1] ... [1,ncols-1]
+//           ...
+//           [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	bufLen := int64(m.mat.Rows*m.mat.Cols*sizeFloat64 + 2*sizeInt64)
+	if bufLen > maxLen {
+		return nil, errDataTooBig
+	}
+
+	buf := bytes.NewBuffer(make([]byte, 0, bufLen))
+	err := binary.Write(buf, binary.LittleEndian, int64(m.mat.Rows))
+	if err != nil {
+		return nil, err
+	}
+	err = binary.Write(buf, binary.LittleEndian, int64(m.mat.Cols))
+	if err != nil {
+		return nil, err
+	}
+
+	nrows, ncols := m.Dims()
+	for i := 0; i < nrows; i++ {
+		for j := 0; j < ncols; j++ {
+			err = binary.Write(buf, binary.LittleEndian, m.at(i, j))
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return buf.Bytes(), nil
+}
+
+// UnmarshalBinary decodes the binary form into the receiver.
+// It panics if the receiver is a non-zero Dense matrix.
+//
+// See MarshalBinary for the on-disk layout.
+//
+// Limited checks on the validity of the binary input are performed:
+//  - matrix.ErrShape is returned if the number of rows or columns is negative,
+//  - an error is returned if the resulting Dense matrix is too
+//  big for the current architecture (e.g. a 16GB matrix written by a
+//  64b application and read back from a 32b application.)
+// UnmarshalBinary does not limit the size of the unmarshaled matrix, and so
+// it should not be used on untrusted data.
+func (m *Dense) UnmarshalBinary(data []byte) error {
+	if !m.isZero() {
+		panic("mat64: unmarshal into non-zero matrix")
+	}
+
+	buf := bytes.NewReader(data)
+	var rows int64
+	err := binary.Read(buf, binary.LittleEndian, &rows)
+	if err != nil {
+		return err
+	}
+	var cols int64
+	err = binary.Read(buf, binary.LittleEndian, &cols)
+	if err != nil {
+		return err
+	}
+
+	m.mat.Rows = int(rows)
+	m.mat.Cols = int(cols)
+	m.mat.Stride = int(cols)
+	m.capRows = int(rows)
+	m.capCols = int(cols)
+
+	size := rows * cols
+	if rows < 0 || cols < 0 {
+		return matrix.ErrShape
+	}
+
+	if int(size) < 0 || size > maxLen {
+		return errDataTooBig
+	}
+
+	m.mat.Data = use(m.mat.Data, int(size))
+
+	return binary.Read(buf, binary.LittleEndian, m.mat.Data)
```

Comment:
Do we want a check to see if the `data []byte` is too long compared to the expected result matrix data?  It won't return an error from `binary.Read`, but it would be an indication that the matrix was not encoded correctly.  See https://play.golang.org/p/k5adVwofft for an example of what I mean.

Ditto with `(v *Vector) UnmarshalBinary(...)`.